IdentityCore/src/throttling/thumbprint_calculation/MSIDThumbprintCalculator.m

```diff
@@ -0,0 +1,104 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.  
+
+
+#import <Foundation/Foundation.h>
+#import "MSIDThumbprintCalculator.h"
+
+
+//Exclude List:
+//1) Client ID - same across all requests
+//2) Grant type - fixed as @"refresh_token"
+@implementation MSIDThumbprintCalculator
+
++ (NSString *)calculateThumbprint:(NSDictionary *)requestParameters
+                     filteringSet:(NSSet *)filteringSet
+                  includePolarity:(BOOL)includePolarity
+{
+    NSArray *sortedThumbprintRequestList = [self sortRequestParametersUsingFilteredSet:requestParameters
+                                                                          filteringSet:filteringSet
+                                                                       includePolarity:includePolarity];
+    if (sortedThumbprintRequestList)
+    {
+        NSUInteger thumbprintKey = [self hash:sortedThumbprintRequestList];
+        if (thumbprintKey == 0)
+        {
+            //Log Warning
+            return nil;
+        }
+        
+        else
+        {
+            return [NSString stringWithFormat:@"%lu", thumbprintKey];
+        }
+    }
+    return nil;
+}
+
++ (NSArray *)sortRequestParametersUsingFilteredSet:(NSDictionary *)requestParameters
+                                      filteringSet:(NSSet *)filteringSet
+                                   includePolarity:(BOOL)includePolarity
+{
+    NSMutableArray *arrayList = [NSMutableArray new];
+    [requestParameters enumerateKeysAndObjectsUsingBlock:^(id  _Nonnull key, id  _Nonnull obj, __unused BOOL * _Nonnull stop) {
+        if ([key isKindOfClass:[NSString class]] && [obj isKindOfClass:[NSString class]])
+        {
+            if ([filteringSet containsObject:key] == includePolarity)
+            {
+                NSArray *thumbprintObject = [NSArray arrayWithObjects:key, obj, nil];
+                [arrayList addObject:thumbprintObject];
+            }
+        }
+    }];
+    
+    NSArray *sortedArrayList = [arrayList sortedArrayUsingComparator:^NSComparisonResult(NSArray *obj1, NSArray *obj2)
+    {
+        return [[obj1 objectAtIndex:0] caseInsensitiveCompare:[obj2 objectAtIndex:0]];
+    }];
+    return sortedArrayList;
+}
+
++ (NSUInteger)hash:(NSArray *)thumbprintRequestList
+{
+    if (!thumbprintRequestList || !thumbprintRequestList.count) return 0;
+    
+    NSUInteger hash = 0;
+    for (id object in thumbprintRequestList)
+    {
+        if ([object isKindOfClass:[NSArray class]] &&
+            ((NSArray *)object).count == 2 &&
+            [object[0] isKindOfClass:[NSString class]])
+        {
+            hash = hash * 31 + ((NSString *)object[1]).hash;
```

Comment:
To make it safer, we can consider hash =(hash * 31) + (key).hash + (value).hash
for example, the 2 requests:
request1 @"key1": "value"
request2 @"key2": "value"
now will have 2 different hash values. What do you think?

Reply:
given that 1) NSString hash always returns the same hash for a given string input 2) keys are invariant, so thus the hash of the keys themselves are invariant. so, it may not offer additional benefits. 
(+ we will also sort them by keys prior to hash calculation.) 
 


To illustrate:  let's say we have two requests with the same request params, with different values for each param: 

Using current hash function, it would produce below outcome, where highlighted fields are hashes of keys. since they are invariant, they don't add additional info to the hash. 

 

Request 1

X : a

Y: b

Z: c

 

 

Request 2

X: b

Y: a

Z: d

 

 

 

Request 1 Hash:  31*(Y + b + 31*(X + a)) + Z + c;

Request 2 Hash:  31*(Y + a + 31*(X + b)) + Z + d;

 

Request 1 Hash :31Y + 31b + 31*31*X + 31*31*a + Z + c;

Request 2 Hash: 31Y + 31a + 31*31*X + 31*31*b + Z + d;